

diff --git a/build/config/BUILD.gn b/build/config/BUILD.gn
index f341687194878..fba9ac5fe0e8c 100644
--- a/build/config/BUILD.gn
+++ b/build/config/BUILD.gn
@@ -171,6 +171,7 @@ config("default_libs") {
       "odbc32.lib",
       "odbccp32.lib",
       "oleaut32.lib",
+      "psapi.lib",
       "shell32.lib",
       "shlwapi.lib",
       "user32.lib",
diff --git a/build/config/win/BUILD.gn b/build/config/win/BUILD.gn
index a20ba08b94da9..650de5854253b 100644
--- a/build/config/win/BUILD.gn
+++ b/build/config/win/BUILD.gn
@@ -243,7 +243,7 @@ config("runtime_library") {
     "_ATL_NO_OPENGL",
     "_WINDOWS",
     "CERT_CHAIN_PARA_HAS_EXTRA_FIELDS",
-    "PSAPI_VERSION=2",
+    "PSAPI_VERSION=1",
     "WIN32",
     "_SECURE_ATL",
   ]
@@ -437,7 +437,6 @@ config("delayloads") {
     "/DELAYLOAD:oleacc.dll",
     "/DELAYLOAD:pdh.dll",
     "/DELAYLOAD:propsys.dll",
-    "/DELAYLOAD:psapi.dll",
     "/DELAYLOAD:rpcrt4.dll",
     "/DELAYLOAD:rstrtmgr.dll",
     "/DELAYLOAD:setupapi.dll",
