diff --git a/components/pdf/renderer/BUILD.gn b/components/pdf/renderer/BUILD.gn
index fb73ef42c6949..651c7be621d79 100644
--- a/components/pdf/renderer/BUILD.gn
+++ b/components/pdf/renderer/BUILD.gn
@@ -38,6 +38,10 @@ static_library("renderer") {
     "//pdf:buildflags",
     "//pdf:features",
     "//pdf:pdf_view_web_plugin",
+    "//ppapi/host",
+    "//ppapi/proxy",
+    "//ppapi/proxy:ipc",
+    "//ppapi/shared_impl",
     "//printing/buildflags",
     "//services/screen_ai/buildflags",
     "//services/strings",
diff --git a/components/pdf/renderer/DEPS b/components/pdf/renderer/DEPS
index 8a20e16eefdae..31597bdcc429a 100644
--- a/components/pdf/renderer/DEPS
+++ b/components/pdf/renderer/DEPS
@@ -11,6 +11,7 @@ include_rules = [
   "+pdf/pdf_accessibility_image_fetcher.h",
   "+pdf/pdf_features.h",
   "+pdf/pdf_view_web_plugin.h",
+  "+ppapi",
   "+printing/buildflags/buildflags.h",
   "+services/screen_ai",
   "+services/strings",
