diff --git a/content/renderer/pepper/pepper_plugin_instance_impl.cc b/content/renderer/pepper/pepper_plugin_instance_impl.cc
index f1a03eb35b33e..3edff2b5f5d41 100644
--- a/content/renderer/pepper/pepper_plugin_instance_impl.cc
+++ b/content/renderer/pepper/pepper_plugin_instance_impl.cc
@@ -2083,6 +2083,7 @@ ppapi::Resource* PepperPluginInstanceImpl::GetSingletonResource(
   // Some APIs aren't implemented in-process.
   switch (id) {
     case ppapi::BROWSER_FONT_SINGLETON_ID:
+    case ppapi::FLASH_FULLSCREEN_SINGLETON_ID:
     case ppapi::ISOLATED_FILESYSTEM_SINGLETON_ID:
     case ppapi::NETWORK_PROXY_SINGLETON_ID:
       NOTIMPLEMENTED();
diff --git a/ppapi/PRESUBMIT.py b/ppapi/PRESUBMIT.py
index fb6bf1b6267f9..e5da57fe851b9 100644
--- a/ppapi/PRESUBMIT.py
+++ b/ppapi/PRESUBMIT.py
@@ -223,6 +223,12 @@ def CheckChange(input_api, output_api):
   # through idl.
   whitelist = ['ppb_opengles2', 'ppb_opengles2ext_dev']
 
+  # The PDF interface is hand-written.
+  # TODO(crbug.com/1302684): Remove the addition below when the APIs they allow
+  # are gone. Removing these entries along with the PDF code in the same CL
+  # confuses this presubmit script.
+  whitelist += ['ppb_pdf', 'ppp_pdf']
+
   # Find all relevant .h and .idl files.
   for filename in files:
     name, ext = os.path.splitext(filename)
diff --git a/ppapi/proxy/ppb_instance_proxy.cc b/ppapi/proxy/ppb_instance_proxy.cc
index 12031dc059cf2..8e8554345d480 100644
--- a/ppapi/proxy/ppb_instance_proxy.cc
+++ b/ppapi/proxy/ppb_instance_proxy.cc
@@ -321,6 +321,9 @@ Resource* PPB_Instance_Proxy::GetSingletonResource(PP_Instance instance,
     case UMA_SINGLETON_ID:
       new_singleton = new UMAPrivateResource(connection, instance);
       break;
+    case FLASH_FULLSCREEN_SINGLETON_ID:
+      NOTREACHED();
+      break;
 // Flash/trusted resources aren't needed for NaCl.
 #if !BUILDFLAG(IS_NACL)
     case BROWSER_FONT_SINGLETON_ID:
diff --git a/ppapi/proxy/ppp_instance_proxy.cc b/ppapi/proxy/ppp_instance_proxy.cc
index 4c25b62373bcd..de545865a4fb9 100644
--- a/ppapi/proxy/ppp_instance_proxy.cc
+++ b/ppapi/proxy/ppp_instance_proxy.cc
@@ -35,6 +35,7 @@ namespace proxy {
 using thunk::EnterInstanceAPINoLock;
 using thunk::EnterInstanceNoLock;
 using thunk::EnterResourceNoLock;
+using thunk::PPB_Flash_Fullscreen_API;
 using thunk::PPB_Instance_API;
 using thunk::PPB_View_API;
 
diff --git a/ppapi/shared_impl/api_id.h b/ppapi/shared_impl/api_id.h
index f9522f0e1f202..e0cfa3136d83c 100644
--- a/ppapi/shared_impl/api_id.h
+++ b/ppapi/shared_impl/api_id.h
@@ -26,6 +26,7 @@ enum ApiID {
   API_ID_PPB_VIDEO_DECODER_DEV,
   API_ID_PPB_X509_CERTIFICATE_PRIVATE,
   API_ID_PPP_CLASS,
+  API_ID_PPP_FIND_PRIVATE,
   API_ID_PPP_GRAPHICS_3D,
   API_ID_PPP_INPUT_EVENT,
   API_ID_PPP_INSTANCE,
diff --git a/ppapi/shared_impl/resource.h b/ppapi/shared_impl/resource.h
index 13dcafe5da2c8..6fb6b70a49325 100644
--- a/ppapi/shared_impl/resource.h
+++ b/ppapi/shared_impl/resource.h
@@ -35,6 +35,8 @@
   F(PPB_FileIO_API)                     \
   F(PPB_FileRef_API)                    \
   F(PPB_FileSystem_API)                 \
+  F(PPB_Find_API)                       \
+  F(PPB_Flash_Fullscreen_API)           \
   F(PPB_Gamepad_API)                    \
   F(PPB_Graphics2D_API)                 \
   F(PPB_Graphics3D_API)                 \
diff --git a/ppapi/shared_impl/singleton_resource_id.h b/ppapi/shared_impl/singleton_resource_id.h
index 36550af2b20f6..6066f725d852d 100644
--- a/ppapi/shared_impl/singleton_resource_id.h
+++ b/ppapi/shared_impl/singleton_resource_id.h
@@ -11,6 +11,7 @@ namespace ppapi {
 // PPB_Instance_API.GetSingletonResource.
 enum SingletonResourceID {
   BROWSER_FONT_SINGLETON_ID,
+  FLASH_FULLSCREEN_SINGLETON_ID,
   GAMEPAD_SINGLETON_ID,
   ISOLATED_FILESYSTEM_SINGLETON_ID,
   NETWORK_PROXY_SINGLETON_ID,
