

diff --git a/skia/BUILD.gn b/skia/BUILD.gn
index 0a64fbdcd52fb..f32da2da03249 100644
--- a/skia/BUILD.gn
+++ b/skia/BUILD.gn
@@ -84,8 +84,6 @@ config("skia_config") {
     "SK_DISABLE_LEGACY_VK_GRDIRECTCONTEXT_FACTORIES",
     "SK_DEFAULT_TYPEFACE_IS_EMPTY",
     "SK_DISABLE_LEGACY_DEFAULT_TYPEFACE",
-    "SK_DISABLE_LEGACY_FONTMGR_REFDEFAULT",
-    "SK_DISABLE_LEGACY_FONTMGR_FACTORY",
   ]
 
   include_dirs = [
