diff --git a/base/win/win_util.cc b/base/win/win_util.cc
index e652023e4c3b1..110bb6c49a24e 100644
--- a/base/win/win_util.cc
+++ b/base/win/win_util.cc
@@ -850,11 +850,17 @@ std::wstring GetWindowObjectName(HANDLE handle) {
 }
 
 bool GetPointerDevice(HANDLE device, POINTER_DEVICE_INFO& result) {
+  if (GetVersion() < Version::WIN8)
+    return false;
+
   return ::GetPointerDevice(device, &result);
 }
 
 std::optional<std::vector<POINTER_DEVICE_INFO>> GetPointerDevices() {
   uint32_t device_count;
+  if (GetVersion() < Version::WIN8)
+    return std::nullopt;
+
   if (!::GetPointerDevices(&device_count, nullptr)) {
     return std::nullopt;
   }
@@ -868,6 +874,9 @@ std::optional<std::vector<POINTER_DEVICE_INFO>> GetPointerDevices() {
 
 bool RegisterPointerDeviceNotifications(HWND hwnd,
                                         bool notify_proximity_changes) {
+  if (GetVersion() < Version::WIN8)
+    return false;
+
   return ::RegisterPointerDeviceNotifications(hwnd, notify_proximity_changes);
 }
 
