

diff --git a/base/hash/md5.h b/base/hash/md5.h
index 215d636fec275..cc831345fc951 100644
--- a/base/hash/md5.h
+++ b/base/hash/md5.h
@@ -64,6 +64,17 @@ BASE_EXPORT std::string MD5DigestToBase16(const MD5Digest& digest);
 // The 'digest' structure will be filled with the result.
 BASE_EXPORT void MD5Sum(base::span<const uint8_t> data, MD5Digest* digest);
 
+// Computes the MD5 sum of the given data buffer with the given length.
+// The given 'digest' structure will be filled with the result data.
+//
+// TODO(https://crbug.com.1490484): Remove this overload, in favor of the one
+// taking `span` (see above).
+BASE_EXPORT inline void MD5Sum(const void* data,
+                               size_t length,
+                               MD5Digest* digest) {
+  MD5Sum(span(static_cast<const uint8_t*>(data), length), digest);
+}
+
 // Returns the MD5 (in hexadecimal) of a string.
 BASE_EXPORT std::string MD5String(std::string_view str);
 
