diff --git a/ppapi/buildflags/buildflags.gni b/ppapi/buildflags/buildflags.gni
index 0bc0e153dff7e..10ce8caaebc6c 100644
--- a/ppapi/buildflags/buildflags.gni
+++ b/ppapi/buildflags/buildflags.gni
@@ -3,7 +3,6 @@
 # found in the LICENSE file.
 
 import("//build/config/cast.gni")
-import("//components/nacl/features.gni")
 
 declare_args() {
   # Enables basic plugin support. This enables support for registering and
@@ -17,9 +16,7 @@ declare_args() {
 
 declare_args() {
   # Enables Pepper API (PPAPI) plugin support.
-  # The only remaining PPAPI plugin supported is NaCl and if is disabled
-  # we can disable PPAPI.
-  enable_ppapi = enable_plugins && enable_nacl && !is_fuchsia
+  enable_ppapi = enable_plugins && !is_fuchsia
 }
 
 assert(enable_plugins || !enable_ppapi)
diff --git a/third_party/blink/web_tests/NeverFixTests b/third_party/blink/web_tests/NeverFixTests
index f892e6d0f42e3..2003d245e6af1 100644
--- a/third_party/blink/web_tests/NeverFixTests
+++ b/third_party/blink/web_tests/NeverFixTests
@@ -267,9 +267,6 @@ crbug.com/627953 [ Mac ] virtual/text-antialias/symbol-cmap.html [ Skip ]
 # these tests can still be run manually with --skipped=ignore.
 webaudio/codec-tests/aac/* [ Skip ]
 
-# PPAPI is disabled on everything but ChromeOS will run via --test-list flag
-ppapi/* [ Skip ]
-
 # WPT for CSS cascade
 crbug.com/763610 external/wpt/css/CSS2/cascade/* [ Skip ]
 
diff --git a/tools/ipc_fuzzer/message_tools/message_list.cc b/tools/ipc_fuzzer/message_tools/message_list.cc
index ee0c2aee26dcf..b9966565e23cd 100644
--- a/tools/ipc_fuzzer/message_tools/message_list.cc
+++ b/tools/ipc_fuzzer/message_tools/message_list.cc
@@ -34,12 +34,7 @@ static msginfo msgtable[] = {
 #include "tools/ipc_fuzzer/message_lib/all_messages.h"
 };
 #define MSGTABLE_SIZE (sizeof(msgtable)/sizeof(msgtable[0]))
-
-#if !BUILDFLAG(ENABLE_NACL) && !BUILDFLAG(ENABLE_PPAPI)
-static_assert(MSGTABLE_SIZE == 0, "There should be no messages");
-#else
 static_assert(MSGTABLE_SIZE, "check your headers for an extra semicolon");
-#endif
 
 static bool check_msgtable() {
   bool result = true;
@@ -57,13 +51,8 @@ static bool check_msgtable() {
 
 #if !BUILDFLAG(ENABLE_NACL)
   exemptions.push_back(NaClMsgStart);
-  exemptions.push_back(NaClHostMsgStart);
 #endif  // !BUILDFLAG(ENABLE_NACL)
 
-#if !BUILDFLAG(ENABLE_PPAPI)
-  exemptions.push_back(PpapiMsgStart);
-#endif  // !BUILDFLAG(ENABLE_PPAPI)
-
   for (size_t i = 0; i < MSGTABLE_SIZE; ++i) {
     int class_id = IPC_MESSAGE_ID_CLASS(msgtable[i].id);
     file_name = msgtable[i].file;
