diff --git a/third_party/adobe/DIR_METADATA b/third_party/adobe/DIR_METADATA
new file mode 100644
index 0000000000000..b8105b6e2d031
--- /dev/null
+++ b/third_party/adobe/DIR_METADATA
@@ -0,0 +1,3 @@
+monorail {
+  component: "Internals>Plugins>Flash"
+}
diff --git a/third_party/adobe/OWNERS b/third_party/adobe/OWNERS
new file mode 100644
index 0000000000000..a6dff391c3c71
--- /dev/null
+++ b/third_party/adobe/OWNERS
@@ -0,0 +1,4 @@
+cpu@chromium.org
+jschuh@chromium.org
+raymes@chromium.org
+viettrungluu@chromium.org
diff --git a/third_party/adobe/README.chromium b/third_party/adobe/README.chromium
new file mode 100644
index 0000000000000..b8b40b04dbb50
--- /dev/null
+++ b/third_party/adobe/README.chromium
@@ -0,0 +1,12 @@
+Name: Adobe Flash
+URL: http://www.adobe.com/products/flashplayer.html
+Version: Varies
+Included In Release: Yes
+Security Critical: Yes
+License: Commercial
+
+Description:
+
+This directory is a glue layer providing targets to allow Chromium to
+build without the Adobe Flash build results necessary to build
+branded Google Chrome.
diff --git a/third_party/adobe/flash/flapper_version.h b/third_party/adobe/flash/flapper_version.h
new file mode 100644
index 0000000000000..c135ad5d52856
--- /dev/null
+++ b/third_party/adobe/flash/flapper_version.h
@@ -0,0 +1,12 @@
+// This is a stand-in for a generated file that is available when Flapper is
+// available.
+
+#ifndef FLAPPER_VERSION_H_INCLUDED_
+#define FLAPPER_VERSION_H_INCLUDED_
+
+// If Flapper were available, we'd define two things:
+//  - FLAPPER_AVAILABLE (to indicate availability), and
+//  - FLAPPER_VERSION_STRING (with the version of the Flapper that's available
+//    as a string, e.g., "11.2.123.456").
+
+#endif  // FLAPPER_VERSION_H_INCLUDED_
