diff --git a/chrome/BUILD.gn b/chrome/BUILD.gn
index 9e7ab3df12290..2b112b2474148 100644
--- a/chrome/BUILD.gn
+++ b/chrome/BUILD.gn
@@ -305,6 +305,13 @@ if (!is_android && !is_mac) {
     # The step's output are needed at runtime, so we also need a data_dep.
     data_deps += [ ":packed_resources" ]
 
+    # Only ChromeOS has precompiled Flash that needs to get copied to the output
+    # directory. On other platforms, Flash is either component-updated only or
+    # not supported at all.
+    if (is_chromeos) {
+      deps += [ "//third_party/adobe/flash:flapper_binaries" ]
+    }
+
     # ChromeOS by design is safe to have rpath=$ORIGIN. This simplifies shared
     # library usage.
     if (is_chromeos && !is_component_build) {
diff --git a/chrome/browser/BUILD.gn b/chrome/browser/BUILD.gn
index ff8ba131edc07..7988e168bf341 100644
--- a/chrome/browser/BUILD.gn
+++ b/chrome/browser/BUILD.gn
@@ -6061,6 +6061,7 @@ static_library("browser") {
     deps += [
       "//ppapi/host",
       "//ppapi/proxy:ipc",
+      "//third_party/adobe/flash:flapper_version_h",
     ]
     if (is_chromeos_ash) {
       deps += [ "//chromeos/ash/components/cryptohome" ]
diff --git a/chrome/browser/ui/BUILD.gn b/chrome/browser/ui/BUILD.gn
index 2c87d77feae16..3416ed95f6bba 100644
--- a/chrome/browser/ui/BUILD.gn
+++ b/chrome/browser/ui/BUILD.gn
@@ -554,6 +554,7 @@ static_library("ui") {
     "//storage/browser",
     "//storage/common",
     "//third_party/abseil-cpp:absl",
+    "//third_party/adobe/flash:flapper_version_h",
     "//third_party/blink/public:buildflags",
     "//third_party/blink/public:resources",
     "//third_party/blink/public/common",
diff --git a/chrome/common/BUILD.gn b/chrome/common/BUILD.gn
index f907062dc4bb8..bdc743bb09ae4 100644
--- a/chrome/common/BUILD.gn
+++ b/chrome/common/BUILD.gn
@@ -453,6 +453,7 @@ static_library("common") {
       "ppapi_utils.cc",
       "ppapi_utils.h",
     ]
+    public_deps += [ "//third_party/adobe/flash:flapper_version_h" ]
   }
   if (enable_plugins && enable_extensions) {
     sources += [
diff --git a/chrome/common/chrome_content_client.cc b/chrome/common/chrome_content_client.cc
index 1805692ba5120..7323ffe7411fa 100644
--- a/chrome/common/chrome_content_client.cc
+++ b/chrome/common/chrome_content_client.cc
@@ -80,6 +80,7 @@
 
 #if BUILDFLAG(ENABLE_PLUGINS)
 #include "content/public/common/content_plugin_info.h"
+#include "flapper_version.h"  // nogncheck  In SHARED_INTERMEDIATE_DIR.
 #endif
 
 #if BUILDFLAG(ENABLE_PDF)
diff --git a/third_party/adobe/flash/BUILD.gn b/third_party/adobe/flash/BUILD.gn
new file mode 100644
index 0000000000000..271a7ca2258de
--- /dev/null
+++ b/third_party/adobe/flash/BUILD.gn
@@ -0,0 +1,68 @@
+# Copyright 2014 The Chromium Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+import("//build/config/chrome_build.gni")
+import("//build/config/chromeos/ui_mode.gni")
+
+if (is_chrome_branded) {
+  if ((is_linux || is_chromeos_lacros) && current_cpu == "x86") {
+    flapper_version_h_file = "symbols/ppapi/linux/flapper_version.h"
+    flapper_binary_files = [
+      "binaries/ppapi/linux/libpepflashplayer.so",
+      "binaries/ppapi/linux/manifest.json",
+    ]
+  } else if ((is_linux || is_chromeos_lacros) && current_cpu == "x64") {
+    flapper_version_h_file = "symbols/ppapi/linux_x64/flapper_version.h"
+    flapper_binary_files = [
+      "binaries/ppapi/linux_x64/libpepflashplayer.so",
+      "binaries/ppapi/linux_x64/manifest.json",
+    ]
+  } else if (is_mac && current_cpu == "x86") {
+    flapper_version_h_file = "symbols/ppapi/mac/flapper_version.h"
+    flapper_binary_files = [
+      "binaries/ppapi/mac/PepperFlashPlayer.plugin",
+      "binaries/ppapi/mac/manifest.json",
+    ]
+  } else if (is_mac && current_cpu == "x64") {
+    flapper_version_h_file = "symbols/ppapi/mac_64/flapper_version.h"
+    flapper_binary_files = [
+      "binaries/ppapi/mac_64/PepperFlashPlayer.plugin",
+      "binaries/ppapi/mac_64/manifest.json",
+    ]
+  } else if (is_win && current_cpu == "x86") {
+    flapper_version_h_file = "symbols/ppapi/win/flapper_version.h"
+    flapper_binary_files = [
+      "binaries/ppapi/win/pepflashplayer.dll",
+      "binaries/ppapi/win/manifest.json",
+    ]
+  } else if (is_win && current_cpu == "x64") {
+    flapper_version_h_file = "symbols/ppapi/win_x64/flapper_version.h"
+    flapper_binary_files = [
+      "binaries/ppapi/win_x64/pepflashplayer.dll",
+      "binaries/ppapi/win_x64/manifest.json",
+    ]
+  } else {
+    flapper_version_h_file = "flapper_version.h"
+    flapper_binary_files = []
+  }
+} else {
+  flapper_version_h_file = "flapper_version.h"
+  flapper_binary_files = []
+}
+
+copy("flapper_version_h") {
+  sources = [ flapper_version_h_file ]
+  outputs = [ "$root_gen_dir/{{source_file_part}}" ]
+}
+
+if (flapper_binary_files == []) {
+  group("flapper_binaries") {
+    # NOP
+  }
+} else {
+  copy("flapper_binaries") {
+    sources = flapper_binary_files
+    outputs = [ "$root_out_dir/PepperFlash/{{source_file_part}}" ]
+  }
+}
