commit b3cadc43a8bd56e05cfb187e3e2c60be427be7fb
Author: John Doe <jdoe@email.com>
Date:   Sun Feb 11 23:43:26 2024 +0800

    SetThreadInformation SetProcessInformation on win7

diff --git a/base/threading/platform_thread_win.cc b/base/threading/platform_thread_win.cc
index 497d471928dde..7c38a56c4c4bc 100644
--- a/base/threading/platform_thread_win.cc
+++ b/base/threading/platform_thread_win.cc
@@ -438,13 +438,17 @@ void SetCurrentThreadPriority(ThreadType thread_type,
   DPLOG_IF(ERROR, !cpu_priority_success)
       << "Failed to set thread priority to " << desired_priority;
 
+  static const auto set_thread_information_fn =
+      reinterpret_cast<decltype(&::SetThreadInformation)>(::GetProcAddress(
+          ::GetModuleHandle(L"kernel32.dll"), "SetThreadInformation"));
   if (g_background_thread_normal_memory_priority_win &&
-      desired_priority == THREAD_MODE_BACKGROUND_BEGIN) {
+      desired_priority == THREAD_MODE_BACKGROUND_BEGIN &&
+      set_thread_information_fn) {
     // Override the memory priority.
     MEMORY_PRIORITY_INFORMATION memory_priority{.MemoryPriority =
                                                     MEMORY_PRIORITY_NORMAL};
     [[maybe_unused]] const BOOL memory_priority_success =
-        SetThreadInformation(thread_handle, ::ThreadMemoryPriority,
+        set_thread_information_fn(thread_handle, ::ThreadMemoryPriority,
                              &memory_priority, sizeof(memory_priority));
     DPLOG_IF(ERROR, !memory_priority_success)
         << "Set thread memory priority failed.";
