

diff --git a/chrome/browser/ui/views/toolbar/toolbar_action_hover_card_bubble_view.cc b/chrome/browser/ui/views/toolbar/toolbar_action_hover_card_bubble_view.cc
index d5243c572c929..5b1b0ac3f6105 100644
--- a/chrome/browser/ui/views/toolbar/toolbar_action_hover_card_bubble_view.cc
+++ b/chrome/browser/ui/views/toolbar/toolbar_action_hover_card_bubble_view.cc
@@ -152,6 +152,9 @@ ToolbarActionHoverCardBubbleView::ToolbarActionHoverCardBubbleView(
   layout->SetCrossAxisAlignment(views::LayoutAlignment::kStretch);
   layout->SetCollapseMargins(true);
 
+  corner_radius_ = ChromeLayoutProvider::Get()->GetCornerRadiusMetric(
+      views::Emphasis::kHigh);
+
   // Set up content.
   auto create_label = [](int context, int text_style,
                          std::optional<ui::ColorId> color_id,
@@ -211,9 +214,9 @@ ToolbarActionHoverCardBubbleView::ToolbarActionHoverCardBubbleView(
   GetBubbleFrameView()->SetPreferredArrowAdjustment(
       views::BubbleFrameView::PreferredArrowAdjustment::kOffset);
   GetBubbleFrameView()->set_hit_test_transparent(true);
-  GetBubbleFrameView()->SetCornerRadius(
-      ChromeLayoutProvider::Get()->GetCornerRadiusMetric(
-          views::Emphasis::kHigh));
+
+  if (using_rounded_corners())
+    GetBubbleFrameView()->SetCornerRadius(corner_radius_.value());
 
   // Start in the fully "faded-in" position so that whatever text we initially
   // display is visible.
@@ -308,6 +311,17 @@ bool ToolbarActionHoverCardBubbleView::IsPolicyLabelVisible() const {
   return policy_label_->GetVisible();
 }
 
+void ToolbarActionHoverCardBubbleView::OnThemeChanged() {
+  BubbleDialogDelegateView::OnThemeChanged();
+
+  // Bubble closes if the theme changes to the point where the border has to be
+  // regenerated. See crbug.com/1140256
+  if (!using_rounded_corners()) {
+    GetWidget()->Close();
+    return;
+  }
+}
+
 ToolbarActionHoverCardBubbleView::~ToolbarActionHoverCardBubbleView() = default;
 
 BEGIN_METADATA(ToolbarActionHoverCardBubbleView)
diff --git a/chrome/browser/ui/views/toolbar/toolbar_action_hover_card_bubble_view.h b/chrome/browser/ui/views/toolbar/toolbar_action_hover_card_bubble_view.h
index 8ab150e351fe8..da95455f8148f 100644
--- a/chrome/browser/ui/views/toolbar/toolbar_action_hover_card_bubble_view.h
+++ b/chrome/browser/ui/views/toolbar/toolbar_action_hover_card_bubble_view.h
@@ -57,6 +57,11 @@ class ToolbarActionHoverCardBubbleView
   class FadeLabel;
   class FootnoteView;
 
+  bool using_rounded_corners() const { return corner_radius_.has_value(); }
+
+  // views::BubbleDialogDelegateView:
+  void OnThemeChanged() override;
+
   // TODO(emiliapaz): rename to `extension_name_label_`.
   raw_ptr<FadeLabelView> title_label_ = nullptr;
   raw_ptr<FadeLabelView> action_title_label_ = nullptr;
@@ -66,6 +71,8 @@ class ToolbarActionHoverCardBubbleView
 
   raw_ptr<views::Separator> site_access_separator_;
   raw_ptr<views::Separator> policy_separator_;
+
+  std::optional<int> corner_radius_;
 };
 
 #endif  // CHROME_BROWSER_UI_VIEWS_TOOLBAR_TOOLBAR_ACTION_HOVER_CARD_BUBBLE_VIEW_H_
