diff --git a/media/base/win/color_space_util_win.cc b/media/base/win/color_space_util_win.cc
index a1f3d150e0242..fbd0816cd6f0d 100644
--- a/media/base/win/color_space_util_win.cc
+++ b/media/base/win/color_space_util_win.cc
@@ -36,8 +36,6 @@ gfx::ColorSpace::PrimaryID MFPrimaryToColorSpace(uint32_t mf_primary) {
       return gfx::ColorSpace::PrimaryID::SMPTEST428_1;
     case MFVideoPrimaries_DCI_P3:
       return gfx::ColorSpace::PrimaryID::SMPTEST431_2;
-    case MFVideoPrimaries_Display_P3:
-      return gfx::ColorSpace::PrimaryID::P3;
     default:
       return gfx::ColorSpace::PrimaryID::INVALID;
   }
@@ -63,8 +61,6 @@ MFVideoPrimaries ColorSpaceToMFPrimary(gfx::ColorSpace::PrimaryID color_space) {
       return MFVideoPrimaries_XYZ;
     case gfx::ColorSpace::PrimaryID::SMPTEST431_2:
       return MFVideoPrimaries_DCI_P3;
-    case gfx::ColorSpace::PrimaryID::P3:
-      return MFVideoPrimaries_Display_P3;
     default:
       return MFVideoPrimaries_Unknown;
   }
@@ -90,10 +86,6 @@ gfx::ColorSpace::TransferID MFTransferToColorSpace(uint32_t mf_transfer) {
       return gfx::ColorSpace::TransferID::PQ;
     case MFVideoTransFunc_HLG:
       return gfx::ColorSpace::TransferID::HLG;
-    case MFVideoTransFunc_BT1361_ECG:
-      return gfx::ColorSpace::TransferID::BT1361_ECG;
-    case MFVideoTransFunc_SMPTE428:
-      return gfx::ColorSpace::TransferID::SMPTEST428_1;
     default:
       return gfx::ColorSpace::TransferID::INVALID;
   }
@@ -121,10 +113,6 @@ MFVideoTransferFunction ColorSpaceToMFTransfer(
       return MFVideoTransFunc_2084;
     case gfx::ColorSpace::TransferID::HLG:
       return MFVideoTransFunc_HLG;
-    case gfx::ColorSpace::TransferID::BT1361_ECG:
-      return MFVideoTransFunc_BT1361_ECG;
-    case gfx::ColorSpace::TransferID::SMPTEST428_1:
-      return MFVideoTransFunc_SMPTE428;
     default:
       return MFVideoTransFunc_Unknown;
   }
@@ -141,14 +129,6 @@ gfx::ColorSpace::MatrixID MFMatrixToColorSpace(uint32_t mf_matrix) {
     case MFVideoTransferMatrix_BT2020_10:
     case MFVideoTransferMatrix_BT2020_12:
       return gfx::ColorSpace::MatrixID::BT2020_NCL;
-    case MFVideoTransferMatrix_Identity:
-      return gfx::ColorSpace::MatrixID::GBR;
-    case MFVideoTransferMatrix_FCC47:
-      return gfx::ColorSpace::MatrixID::FCC;
-    case MFVideoTransferMatrix_YCgCo:
-      return gfx::ColorSpace::MatrixID::YCOCG;
-    case MFVideoTransferMatrix_SMPTE2085:
-      return gfx::ColorSpace::MatrixID::YDZDX;
     default:
       return gfx::ColorSpace::MatrixID::INVALID;
   }
@@ -165,14 +145,6 @@ MFVideoTransferMatrix ColorSpaceToMFMatrix(
       return MFVideoTransferMatrix_SMPTE240M;
     case gfx::ColorSpace::MatrixID::BT2020_NCL:
       return MFVideoTransferMatrix_BT2020_10;
-    case gfx::ColorSpace::MatrixID::GBR:
-      return MFVideoTransferMatrix_Identity;
-    case gfx::ColorSpace::MatrixID::FCC:
-      return MFVideoTransferMatrix_FCC47;
-    case gfx::ColorSpace::MatrixID::YCOCG:
-      return MFVideoTransferMatrix_YCgCo;
-    case gfx::ColorSpace::MatrixID::YDZDX:
-      return MFVideoTransferMatrix_SMPTE2085;
     default:
       return MFVideoTransferMatrix_Unknown;
   }
