

diff --git a/chrome/app/chromium_strings.grd b/chrome/app/chromium_strings.grd
index 4df298c132c9f..b594b8f9ae1a4 100644
--- a/chrome/app/chromium_strings.grd
+++ b/chrome/app/chromium_strings.grd
@@ -625,7 +625,7 @@ Permissions you've already given to websites and apps may apply to this account.
           Can not install the same Chromium version that is currently running. Please close Chromium and try again.
         </message>
         <message name="IDS_INSTALL_OS_NOT_SUPPORTED" desc="Error displayed if OS is not supported">
-          Chromium requires Windows 10 or higher.
+          Chromium requires Windows 7 or higher.
         </message>
         <message name="IDS_INSTALL_OS_ERROR" desc="Error displayed when any Windows API call fails and we do not have more specific information.">
           An operating system error occurred during installation. Please download Chromium again.
diff --git a/chrome/app/google_chrome_strings.grd b/chrome/app/google_chrome_strings.grd
index 75ca16d39adb3..99cc91f11eab3 100644
--- a/chrome/app/google_chrome_strings.grd
+++ b/chrome/app/google_chrome_strings.grd
@@ -602,7 +602,7 @@ Permissions you've already given to websites and apps may apply to this account.
           Can not install the same Google Chrome version that is currently running. Please close Google Chrome and try again.
         </message>
         <message name="IDS_INSTALL_OS_NOT_SUPPORTED" desc="Error displayed if OS is not supported">
-          Google Chrome requires Windows 10 or higher.
+          Google Chrome requires Windows 7 or higher.
         </message>
         <message name="IDS_INSTALL_OS_ERROR" desc="Error displayed when any Windows API call fails and we do not have more specific information.">
           An operating system error occurred during installation. Please download Google Chrome again.
