

diff --git a/mojo/core/embedder/embedder.cc b/mojo/core/embedder/embedder.cc
index 44651b630f13a..591c148af8b4c 100644
--- a/mojo/core/embedder/embedder.cc
+++ b/mojo/core/embedder/embedder.cc
@@ -42,12 +42,7 @@ namespace mojo::core {
 
 namespace {
 
-#if BUILDFLAG(IS_CHROMEOS) || BUILDFLAG(IS_FUCHSIA)
 std::atomic<bool> g_mojo_ipcz_enabled{false};
-#else
-// Default to enabled even if InitFeatures() is never called.
-std::atomic<bool> g_mojo_ipcz_enabled{true};
-#endif
 
 std::optional<std::string> GetMojoIpczEnvVar() {
   std::string value;
diff --git a/mojo/core/embedder/features.cc b/mojo/core/embedder/features.cc
index a3b2ae1b54986..8c788a6581e66 100644
--- a/mojo/core/embedder/features.cc
+++ b/mojo/core/embedder/features.cc
@@ -33,11 +33,7 @@ BASE_FEATURE(kMojoInlineMessagePayloads,
              "MojoInlineMessagePayloads",
              base::FEATURE_DISABLED_BY_DEFAULT);
 
-#if BUILDFLAG(IS_CHROMEOS) || BUILDFLAG(IS_FUCHSIA)
 BASE_FEATURE(kMojoIpcz, "MojoIpcz", base::FEATURE_DISABLED_BY_DEFAULT);
-#else
-BASE_FEATURE(kMojoIpcz, "MojoIpcz", base::FEATURE_ENABLED_BY_DEFAULT);
-#endif
 
 }  // namespace core
 }  // namespace mojo
