

diff --git a/mojo/core/embedder/embedder.cc b/mojo/core/embedder/embedder.cc
index dca56dd4df86e..de999ee3c7d64 100644
--- a/mojo/core/embedder/embedder.cc
+++ b/mojo/core/embedder/embedder.cc
@@ -43,10 +43,5 @@ namespace mojo::core {
 namespace {
 
 #if BUILDFLAG(MOJO_SUPPORT_LEGACY_CORE)
-#if BUILDFLAG(IS_CHROMEOS) && !defined(ENABLE_IPCZ_ON_CHROMEOS)
 std::atomic<bool> g_mojo_ipcz_enabled{false};
-#else
-// Default to enabled even if InitFeatures() is never called.
-std::atomic<bool> g_mojo_ipcz_enabled{true};
-#endif
 
diff --git a/mojo/core/embedder/features.cc b/mojo/core/embedder/features.cc
index a3b2ae1b54986..8c788a6581e66 100644
--- a/mojo/core/embedder/features.cc
+++ b/mojo/core/embedder/features.cc
@@ -33,11 +33,7 @@ BASE_FEATURE(kMojoInlineMessagePayloads,
 BASE_FEATURE(kMojoInlineMessagePayloads, base::FEATURE_DISABLED_BY_DEFAULT);
 
 #if BUILDFLAG(MOJO_SUPPORT_LEGACY_CORE)
-#if BUILDFLAG(IS_CHROMEOS)
 BASE_FEATURE(kMojoIpcz, base::FEATURE_DISABLED_BY_DEFAULT);
-#else
-BASE_FEATURE(kMojoIpcz, base::FEATURE_ENABLED_BY_DEFAULT);
-#endif
 #endif  // BUILDFLAG(MOJO_SUPPORT_LEGACY_CORE)
 
 BASE_FEATURE(kMojoIpczMemV2, base::FEATURE_DISABLED_BY_DEFAULT);
diff --git a/mojo/features.gni b/mojo/features.gni
index ce164c3cd36ec..f54981fcb54c3 100644
--- a/mojo/features.gni
+++ b/mojo/features.gni
@@ -18,5 +18,5 @@ declare_args() {
   #
   # We also enable this by default for fuzzer builds, since some active fuzzers
   # require the legacy implementation.
-  mojo_support_legacy_core = target_os == "chromeos" || use_libfuzzer
+  mojo_support_legacy_core = target_os == "chromeos" || use_libfuzzer || is_win
 }
