diff --git a/chrome/installer/util/taskbar_util.cc b/chrome/installer/util/taskbar_util.cc
index 8909aaddb98fd..0329447fdd3ad 100644
--- a/chrome/installer/util/taskbar_util.cc
+++ b/chrome/installer/util/taskbar_util.cc
@@ -116,13 +116,33 @@ bool CanPinShortcutToTaskbar() {
   if (g_can_pin_to_taskbar && !(*g_can_pin_to_taskbar)()) {
     return false;
   }
-  return PinShortcutWithIPinnedList3(shortcut);
+  base::ScopedBlockingCall scoped_blocking_call(FROM_HERE,
+                                                base::BlockingType::MAY_BLOCK);
+  DCHECK(CanPinShortcutToTaskbar());
+  if (base::win::GetVersion() >= base::win::Version::WIN10_RS5)
+    return PinShortcutWithIPinnedList3(shortcut);
+
+  intptr_t result = reinterpret_cast<intptr_t>(ShellExecute(
+      nullptr, L"taskbarpin", shortcut.value().c_str(), nullptr, nullptr, 0));
+  return result > 32;
 }
 
 bool UnpinShortcutFromTaskbar(const base::FilePath& shortcut) {
-  return UnpinShortcutWithIPinnedList3(shortcut);
+  base::ScopedBlockingCall scoped_blocking_call(FROM_HERE,
+                                                base::BlockingType::MAY_BLOCK);
+  // Calling ShellExecute can be crashy because of shell hooks/malware, so try
+  // using IPinnedList3. Fallback to ShellExecute if it fails.
+  if (base::win::GetVersion() >= base::win::Version::WIN10_RS5 &&
+      UnpinShortcutWithIPinnedList3(shortcut)) {
+    return true;
+  }
+
+  intptr_t result = reinterpret_cast<intptr_t>(ShellExecute(
+      nullptr, L"taskbarunpin", shortcut.value().c_str(), nullptr, nullptr, 0));
+  return result > 32;
 }
 
+// static
 std::optional<bool> IsShortcutPinnedToTaskbar(const base::FilePath& shortcut) {
   Microsoft::WRL::ComPtr<IPinnedList3> pinned_list = GetTaskbarPinnedList();
   if (!pinned_list.Get())
diff --git a/chrome/installer/util/taskbar_util.h b/chrome/installer/util/taskbar_util.h
index 650295c5c975a..76e8d9966367d 100644
--- a/chrome/installer/util/taskbar_util.h
+++ b/chrome/installer/util/taskbar_util.h
@@ -14,12 +14,13 @@ namespace base {
 class FilePath;
 }  // namespace base
 
-// Pin to taskbar is supported on Win10RS5+. Returns true on those platforms.
+// Pin to taskbar is supported on Windows 7, 8 and Win10RS5+. Returns
+// true on those platforms.
 bool CanPinShortcutToTaskbar();
 
-// Pins a shortcut to the taskbar on supported platforms. The `shortcut` file
-// must already exist and be a shortcut that points to an executable. The app id
-// of the shortcut is used to group windows and must be set correctly.
+// Pins a shortcut to the taskbar on Windows 7, 8 and Win10RS5+ . `shortcut`
+// file must already exist and be a shortcut that points to an executable.The
+// app id of the shortcut is used to group windows and must be set correctly.
 bool PinShortcutToTaskbar(const base::FilePath& shortcut);
 
 // Unpins a shortcut from the Windows 7+ taskbar. `shortcut` must exist
