diff --git a/modules/desktop_capture/desktop_capture_options.h b/modules/desktop_capture/desktop_capture_options.h
index bc1342a648..c62476f405 100644
--- a/modules/desktop_capture/desktop_capture_options.h
+++ b/modules/desktop_capture/desktop_capture_options.h
@@ -213,20 +213,6 @@ class RTC_EXPORT DesktopCaptureOptions {
   // The flag has no effect if the allow_wgc_capturer flag is false.
   bool wgc_require_border() const { return wgc_require_border_; }
   void set_wgc_require_border(bool require) { wgc_require_border_ = require; }
-
-  // For window capture, set to true to include more application content like
-  // tool tips and drop downs. From the Microsoft developer docs:
-  //
-  // "Secondary Windows are considered to be windows that have either the
-  // WS_POPUP or WS_EX_TOOLWINDOW styles that intersect the main window. The
-  // windows are drawn into the texture the app receives and are clipped if they
-  // go outside the bounds of the main top level window."
-  bool wgc_include_secondary_windows() const {
-    return wgc_include_secondary_windows_;
-  }
-  void set_wgc_include_secondary_windows(bool include) {
-    wgc_include_secondary_windows_ = include;
-  }
 #endif  // defined(RTC_ENABLE_WIN_WGC)
 #endif  // defined(WEBRTC_WIN)
 
@@ -284,7 +270,6 @@ class RTC_EXPORT DesktopCaptureOptions {
   bool allow_wgc_capturer_fallback_ = false;
   bool allow_wgc_zero_hertz_ = false;
   bool wgc_require_border_ = false;
-  bool wgc_include_secondary_windows_ = false;
 #endif
 #endif
 #if defined(WEBRTC_USE_X11)
diff --git a/modules/desktop_capture/win/wgc_capture_session.cc b/modules/desktop_capture/win/wgc_capture_session.cc
index 3969654bce..01ec96b6c1 100644
--- a/modules/desktop_capture/win/wgc_capture_session.cc
+++ b/modules/desktop_capture/win/wgc_capture_session.cc
@@ -270,18 +270,6 @@ HRESULT WgcCaptureSession::StartCapture(const DesktopCaptureOptions& options) {
     session3->put_IsBorderRequired(options.wgc_require_border());
   }
 
-  // Windows 11 24H2 (10.0.26100.0) added
-  // `IGraphicsCaptureSession6::put_IncludeSecondaryWindows()`. See
-  // `wgc_include_secondary_windows()` in
-  // /modules/desktop_capture/desktop_capture_options.h for more details.
-  ComPtr<ABI::Windows::Graphics::Capture::IGraphicsCaptureSession6> session6;
-  if (SUCCEEDED(session_->QueryInterface(
-          ABI::Windows::Graphics::Capture::IID_IGraphicsCaptureSession6,
-          &session6))) {
-    session6->put_IncludeSecondaryWindows(
-        options.wgc_include_secondary_windows());
-  }
-
   allow_zero_hertz_ = options.allow_wgc_zero_hertz();
 
   hr = session_->StartCapture();
