diff --git a/modules/desktop_capture/win/dxgi_output_duplicator.cc b/modules/desktop_capture/win/dxgi_output_duplicator.cc
index 16299c01fe..ed0601c2c9 100644
--- a/modules/desktop_capture/win/dxgi_output_duplicator.cc
+++ b/modules/desktop_capture/win/dxgi_output_duplicator.cc
@@ -73,9 +73,41 @@ DxgiOutputDuplicator::DxgiOutputDuplicator(const D3dDevice& device,
   RTC_DCHECK(!desktop_rect_.is_empty());
   RTC_DCHECK_GT(desktop_rect_.width(), 0);
   RTC_DCHECK_GT(desktop_rect_.height(), 0);
-  HRESULT hr = GetScaleFactorForMonitor(desc.Monitor, &device_scale_factor_);
-  RTC_LOG_IF(LS_ERROR, FAILED(hr))
-      << "Failed to get scale factor for monitor: " << hr;
+#define GETPERCENT(dpi) ((dpi * 100 + 50) / 96)
+typedef HRESULT (* LPFNDLLFUNC1)(HMONITOR,DEVICE_SCALE_FACTOR *);
+HINSTANCE hDLL;               // Handle to DLL
+LPFNDLLFUNC1 lpfnDllFunc1;    // Function pointer
+hDLL = LoadLibrary(TEXT("Shcore.dll"));
+if (hDLL != NULL)
+{
+   lpfnDllFunc1 = (LPFNDLLFUNC1)GetProcAddress(hDLL,
+                                           "GetScaleFactorForMonitor");
+   if (!lpfnDllFunc1)
+   {
+      // handle the error
+      FreeLibrary(hDLL);
+        //---- set screen dpi (per session) ----
+        HDC hdc = GetWindowDC(NULL);
+        if (! hdc)
+        {
+           RTC_LOG(LS_ERROR)
+              << "Failed to get scale factor for monitor: ";
+        }
+        else
+        {
+            int iDpi = GetDeviceCaps(hdc, LOGPIXELSX);
+            ReleaseDC(NULL, hdc);
+            device_scale_factor_ = (DEVICE_SCALE_FACTOR) GETPERCENT(iDpi);
+        }
+   }
+   else
+   {
+      // call the function
+      HRESULT hr = lpfnDllFunc1(desc.Monitor, &device_scale_factor_);
+      RTC_LOG_IF(LS_ERROR, FAILED(hr))
+          << "Failed to get scale factor for monitor: " << hr;
+   }
+}
 }
 
 DxgiOutputDuplicator::DxgiOutputDuplicator(DxgiOutputDuplicator&& other) =
diff --git a/modules/desktop_capture/win/wgc_capture_session.cc b/modules/desktop_capture/win/wgc_capture_session.cc
index 0d4b33752a..e33eb4bc98 100644
--- a/modules/desktop_capture/win/wgc_capture_session.cc
+++ b/modules/desktop_capture/win/wgc_capture_session.cc
@@ -113,9 +113,41 @@ WgcCaptureSession::WgcCaptureSession(intptr_t source_id,
     monitor = MonitorFromWindow(reinterpret_cast<HWND>(source_id),
                                 /*dwFlags=*/MONITOR_DEFAULTTONEAREST);
   }
-  HRESULT hr = GetScaleFactorForMonitor(monitor, &device_scale_factor_);
-  RTC_LOG_IF(LS_ERROR, FAILED(hr))
-      << "Failed to get scale factor for monitor: " << hr;
+#define GETPERCENT(dpi) ((dpi * 100 + 50) / 96)
+typedef HRESULT (* LPFNDLLFUNC1)(HMONITOR,DEVICE_SCALE_FACTOR *);
+HINSTANCE hDLL;               // Handle to DLL
+LPFNDLLFUNC1 lpfnDllFunc1;    // Function pointer
+hDLL = LoadLibrary(TEXT("Shcore.dll"));
+if (hDLL != NULL)
+{
+   lpfnDllFunc1 = (LPFNDLLFUNC1)GetProcAddress(hDLL,
+                                           "GetScaleFactorForMonitor");
+   if (!lpfnDllFunc1)
+   {
+      // handle the error
+      FreeLibrary(hDLL);
+        //---- set screen dpi (per session) ----
+        HDC hdc = GetWindowDC(NULL);
+        if (! hdc)
+        {
+           RTC_LOG(LS_ERROR)
+              << "Failed to get scale factor for monitor: ";
+        }
+        else
+        {
+            int iDpi = GetDeviceCaps(hdc, LOGPIXELSX);
+            ReleaseDC(NULL, hdc);
+            device_scale_factor_ = (DEVICE_SCALE_FACTOR) GETPERCENT(iDpi);
+        }
+   }
+   else
+   {
+      // call the function
+      HRESULT hr = GetScaleFactorForMonitor(monitor, &device_scale_factor_);
+      RTC_LOG_IF(LS_ERROR, FAILED(hr))
+          << "Failed to get scale factor for monitor: " << hr;
+   }
+}
 }
 
 WgcCaptureSession::~WgcCaptureSession() {
